### ISSUE TITLE
Consider creating a tablesaw-ml module

### ISSUE THREAD

**Issue by agibsonccc** [jtablesaw/tablesaw#160]
Hey folks,

We will have to get rid of tablesaw if this smile dep on the core continues. The tablesaw dep is a bit heavy with smile built right in.
Can we maybe have the table be in a "core library" and relevant smile bits in a submodule?

**Comment by benmccann** [comment]
We did pretty substantial work moving dependencies out of core recently such as creating a separate plot module. I had considered creating something like a tablesaw-ml module. However, smile is a very lightweight dependency with no transitive dependencies, so I'm not sure it'd make much of a difference. Can you expand on what effect removing smile would have for you?

**Comment by lwhite1** [comment]
Hi Adam,

Sorry to hear this is an issue for you. We'll try to accommodate as much as possible. 

From our perspective, it would be ideal to package ML as a separate project, as that would let us plug-in other ML libraries like SparkML and H2O. 

At this point, however, our Smile interface is lacking so many elements that I think we would prefer to prioritize building it out so that we could offer at least one "full" solution to our users. For me, full means you could use Tablesaw in a Kaggle competition without being disadvantaged to the people using R or Python.

To the best of my knowledge, there isn't much overlap between what Smile offers and what DeepLearning4J offers. For that reason, I've been interested for some time in adding a DeepLearning module that would give Tablesaw users easy access to your library/ecosystem. Again, thinking of our users, this is something we'd prefer to prioritize over removing Smile.

Anyway, I look forward to hearing more about your concerns.

**Comment by agibsonccc** [comment]
Right so I don't think our goals are incompatible here. You could more or less even keep backwards compatibility by just keeping the same namespaces but refactoring out the separate module.
One of the problems I'm going to run in to depending on table saw is smile depends on nd4j, our matrix library. I can't have an incompatible nd4j dep on our classpath. 

The size is the main issue though..we start including ML, plotting,.. when all we need is the basic columnar processing, we end up with a big conflict of interest here.

This is a big reason why I forked in the first place. We only needed the columnar backend originally.

If our goals don't align that's fine, but overall  I think I'll just be in favor of dropping tablesaw. We are switching to a more binary approach anyways. I'll close this for now and wish you the best of luck with the lib.

**Comment by lwhite1** [comment]
We want this open as an issue regardless of the DeepLearning4J's ultimate determination on using Tablesaw. 

@agibsonccc please note that we've already removed plotting from the core library

**Comment by benmccann** [comment]
@agibsonccc we only depend on smile-core, so there shouldn't be any dependency on nd4j

were still aligned on minimizing dependencies. this has been a big focus for us recently and we'll continue to work on it. let me know if i'm wrong about nd4j and we can look at prioritizing that one

**Comment by benmccann** [comment]
I'm wondering if we add a few more methods to make it easier to create the types that Smile expects if we need to wrap all of their classes actually. It'd be kind of nice if we could avoid this since then we could support their entire library instead of just the subset we support now. I filed a new issue to consider that question https://github.com/jtablesaw/tablesaw/issues/168

**Comment by benmccann** [comment]
I've sent a PR for this: https://github.com/jtablesaw/tablesaw/pull/174

**Comment by lwhite1** [comment]
Fixed in #174.  closing

**Comment by agibsonccc** [comment]
Thanks guys!
